### PR TITLE
readme: explain co-operation between cbs and bulkheads

### PR DESCRIPTION
**Pull request by sirupsen**
On Friday I was explaining Semian to @dalehamel, @dwradcliffe and @kirs. The question came up about when circuit breakers are useful and when bulkheads are. To explain the co-operation, we ended up drawing the graphs below on a whiteboard to explain how they work together. This explanation seemed like a useful addition to the README.

resiliency- circuit breakers 
resiliency- bulkheads
resiliency- circuit breakers bulkheads

@Shopify/resiliency @BoGs @jacobbednarz

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -354,24 +354,25 @@ There are three configuration parameters for circuit breakers in Semian:
 
 ### Bulkheading
 
-For many applications, circuit breakers are not enough however. This is best
-illustrated with an extreme. Imagine if the timeout for our data store isn't as
-low as 200ms, but actually 10 seconds. For example, you might have a relational data
+For some applications, circuit breakers are not enough.j This is best illustrated
```
Comment (jacobbednarz) [inline README.md]: Looks like you've got a rogue `j` in here.

**Comment by jacobbednarz** [comment]
👍  great addition

File: README.md [hunk 2]
```diff
@@ -354,24 +354,25 @@ There are three configuration parameters for circuit breakers in Semian:
...
+with an extreme. Imagine if the timeout for our data store isn't as low as
```
Comment (camilo) [inline README.md]: "with an extreme" feels to me it needs a noun. It could be "with an extreme example".

File: README.md [hunk 3]
```diff
@@ -433,6 +434,35 @@ against a misbehaving resource. The driver will raise an exception after trying
 to connect with a timeout or after an immediate failure. These driver actions
 will affect the circuit and Semian, which can make future calls fail faster.
 
+A useful way to think about the co-operation between bulkheads and circuit
+breakers is through visualizing a failure scenario graphing capacity as a
+function of time. If an incident strikes that makes the server unresponsive with
```
Comment (camilo) [inline README.md]: I might be ESLing here but would " incident strikes which makes" be better?

File: README.md [hunk 4]
```diff
@@ -433,6 +434,35 @@ against a misbehaving resource. The driver will raise an exception after trying
...
+a `20s` timeout on the client and you only have circuit breakers enabled--you
+will loose capacity until all workers have tripped their circuit breakers. The
+slope of this line will depend on the amount of traffic to the now unavailable
+service. If the slope is steep (i.e. high traffic), you'll loose capacity
```
Comment (camilo) [inline README.md]: s/loose/lose/

File: README.md [hunk 5]
```diff
@@ -433,6 +434,35 @@ against a misbehaving resource. The driver will raise an exception after trying
...
+quicker. The higher the client driver timeout, the longer you'll loose capacity
```
Comment (camilo) [inline README.md]: same here lose vs loose

File: README.md [hunk 6]
```diff
@@ -433,6 +434,35 @@ against a misbehaving resource. The driver will raise an exception after trying
...
+![resiliency- circuit breakers](https://cloud.githubusercontent.com/assets/97400/22405538/53229758-e612-11e6-81b2-824f873c3fb7.png)
+
+If we imagine the same scenario but with _only_ bulkheads, configured to have
+tickets for 50% of workers at any given time, we'll see the following
+flat-lining scenario:
+
+![resiliency- bulkheads](https://cloud.githubusercontent.com/assets/97400/22405542/6832a372-e612-11e6-88c4-2452b64b3121.png)
+
+Circuit breakers have the nice property of re-gaining 100% capacity. Bulkheads
+have the desirable property of guaranteeing a minimum capacity. If we do
```
Comment (csfrancis) [inline README.md]: Do you mean maximum capacity here? The way I envision bulkheads is that they define the maximum number of users (tickets) for a target resource.

Comment (sirupsen) [inline README.md]: Yeah, it's whether you see the glass half full, or half empty. You can think of Semian as providing minimum capacity for everything else, or a maximum capacity for a single resource.

**Pull request merged**